test/main/assert/timeout.js

```diff
@@ -0,0 +1,61 @@
+QUnit.module( "assert.timeout" );
+
+QUnit.test( "pushes a failure if test times out when using async", function( assert ) {
+	assert.timeout( 10 );
+	assert.expect( 1 );
+
+	var originalPushFailure = QUnit.config.current.pushFailure;
+	QUnit.config.current.pushFailure = function pushFailureStub( message ) {
+		QUnit.config.current.pushFailure = originalPushFailure;
+
+		assert.equal( message, "Test took longer than 10ms; test timed out." );
+	};
+
+	assert.async();
+} );
+
+QUnit.test( "pushes a failure if test times out when using a promise", function( assert ) {
+	assert.timeout( 10 );
+	assert.expect( 1 );
+
+	var originalPushFailure = QUnit.config.current.pushFailure;
+	QUnit.config.current.pushFailure = function pushFailureStub( message ) {
+		QUnit.config.current.pushFailure = originalPushFailure;
+
+		assert.equal( message, "Test took longer than 10ms; test timed out." );
+	};
+
+	// Return a "thenable" to serve as a mock Promise
+	return {
+		then: function() {}
+	};
+} );
+
+QUnit.test( "does not push a failure if test is synchronous", function( assert ) {
+	assert.timeout( 1 );
+
+	var wait = Date.now() + 50;
```

Comment:
Let's use `10` here as with the other one. Will make the test run a bit faster, too.